src/OrleansRuntime/Catalog/IGrainActivationContext.cs

```diff
@@ -0,0 +1,20 @@
+﻿using System;
+using Orleans.Core;
+
+namespace Orleans.Runtime
+{
+    /// <summary>
+    /// The context associated with the current grain activation.
+    /// </summary>
+    public interface IGrainActivationContext
+    {
+        /// <summary>Gets the .NET type of the grain activation instance.</summary>
+        Type GrainType { get; }
+
+        /// <summary>Gets the identity of the grain activation.</summary>
+        IGrainIdentity GrainIdentity { get; }
+
+        /// <summary>Gets the <see cref="IServiceProvider"/> that provides access to the grain's service container.</summary>
+        IServiceProvider GrainServices { get; }
```

Comment:
well, that's the point, it's not the same as ApplicationServices (actually it currently is the same instance, but this will definitely change once we add scoped services in #2856). This will have the IServiceProvider scoped to this particular grain activation, similar to MVC's context.RequestServices, which is a scoped instance separate from their ApplicationServices.